catalog/core/catalog-core-transform-impl/src/main/java/org/codice/ddf/catalog/locator/impl/TransformerLocatorImpl.java

```diff
@@ -0,0 +1,263 @@
+
+    return null;
+  }
+
+  private Map<String, Object> getServiceProperties(ServiceReference serviceReference) {
+    return Arrays.stream(serviceReference.getPropertyKeys())
+        .collect(Collectors.toMap(s -> s, serviceReference::getProperty));
+  }
+
+  protected BundleContext getContext() {
+    Bundle bundle = FrameworkUtil.getBundle(TransformerLocatorImpl.class);
+    if (bundle != null) {
+      return bundle.getBundleContext();
+    }
+    return null;
+  }
+
+  private <T> Collection<ServiceReference<T>> findServices(
+      Class<T> clazz, String filter, BundleContext bundleContext) {
+    try {
+      return bundleContext.getServiceReferences(clazz, filter);
+    } catch (InvalidSyntaxException e) {
+      throw new IllegalArgumentException("Invalid syntax supplied");
+    }
+  }
+
+  private <T extends TransformerProperties> List<T> findTransformers(
+      Class<T> clazz, String filter, Predicate<T> transformerPropertiesFilter) {
+    BundleContext bundleContext = getContext();
+
+    Collection<ServiceReference<T>> refs = findServices(clazz, filter, bundleContext);
+
+    if (refs != null) {
+      return refs.stream()
+          .sorted(Collections.reverseOrder())
+          .map(bundleContext::getService)
+          .filter(transformerPropertiesFilter)
+          .collect(Collectors.toList());
+    }
+
+    return Collections.emptyList();
+  }
+
+  private <T, R> List<R> findTransformers(
+      Class<T> clazz,
+      BiFunction<T, Map<String, Object>, R> converter,
+      String filter,
+      Predicate<Map<String, Object>> servicePropertiesFilter) {
+    BundleContext bundleContext = getContext();
+
+    Collection<ServiceReference<T>> refs = findServices(clazz, filter, bundleContext);
+
+    if (refs != null) {
+      return refs.stream()
+          .sorted(Collections.reverseOrder())
+          .map(ref -> new ImmutablePair<>(bundleContext.getService(ref), getServiceProperties(ref)))
+          .filter(pair -> servicePropertiesFilter.test(pair.getRight()))
+          .map(pair -> converter.apply(pair.getLeft(), pair.getRight()))
+          .collect(Collectors.toList());
+    }
+
+    return Collections.emptyList();
+  }
+
+  private <M extends TransformerProperties, S> List<M> findMultiTransformers(
+      Class<M> multiClazz,
+      Class<S> singleClazz,
+      BiFunction<S, Map<String, Object>, M> convertSingleToMulti,
+      String filter,
+      Predicate<M> transformerPropertiesFilter,
+      Predicate<Map<String, Object>> servicePropertiesFilter) {
+
+    List<M> multiMetacardTransformers = new LinkedList<>();
+
+    multiMetacardTransformers.addAll(
+        findTransformers(multiClazz, filter, transformerPropertiesFilter));
+
+    multiMetacardTransformers.addAll(
+        findTransformers(singleClazz, convertSingleToMulti, filter, servicePropertiesFilter));
+
+    return multiMetacardTransformers;
+  }
+
+  private <M extends TransformerProperties, S> List<M> findMultiTransformers(
+      Class<M> multiClazz,
```

Comment:
Is there a reason you can't call these multiClass and singleClass?

Reply:
There's no technical reason. There is the convention to use `clazz` in most cases. So I was being consistent with that convention.